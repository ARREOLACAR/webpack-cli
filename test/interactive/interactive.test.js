'use strict';
const { run, runAndGetWatchProc } = require('../utils/test-utils');
const { stat, writeFileSync } = require('fs');
const { resolve } = require('path');
const { version } = require('webpack');

const wordsInStatsv4 = ['Hash', 'Version', 'Time', 'Built at:', 'main.js'];
const wordsInStatsv5 = ['asset', 'index.js', 'compiled successfully'];

describe('--interactive flag', () => {
    it('should add InteractiveModePlugin to config', (done) => {
        const { stderr, stdout } = run(__dirname, ['--interactive']);
        expect(stderr).toBeFalsy();
        expect(stdout).toContain('InteractiveModePlugin');
        stat(resolve(__dirname, './bin/main.js'), (err, stats) => {
            expect(err).toBe(null);
            expect(stats.isFile()).toBe(true);
            done();
        });
    });

    it('should work with --watch', (done) => {
        const proc = runAndGetWatchProc(__dirname, ['-c', './webpack.watch.js', '--watch', '--interactive'], false, '', true);
        let semaphore = 2;
        const clear = '\x1B[2J\x1B[3J\x1B[H';
        proc.stdout.on('data', (chunk) => {
            const data = chunk.toString();
            if (data.includes('watching files for updates')) {
                writeFileSync(resolve(__dirname, 'index.js'), `console.log('I am Batman');`);
                semaphore--;
                return;
            }
            if (semaphore === 1) {
                expect(data).toBe(clear);
                semaphore--;
                return;
            }
            if (semaphore === 0) {
                if (version.startsWith('5')) {
                    for (const word of wordsInStatsv5) {
                        expect(data).toContain(word);
                    }
                } else {
                    for (const word of wordsInStatsv4) {
                        expect(data).toContain(word);
                    }
                }
                proc.kill();
                done();
                return;
            }
        });
    });

    it('should not output in interactive with --watch only', (done) => {
        const proc = runAndGetWatchProc(__dirname, ['-c', './webpack.watch.js', '--watch'], false, '', true);
        let semaphore = 1;
        proc.stdout.on('data', (chunk) => {
            const data = chunk.toString();
            if (data.includes('watching files for updates')) {
                writeFileSync(resolve(__dirname, 'index.js'), `console.log('I am Batman');`);
                semaphore--;
                return;
            }
            if (semaphore === 0) {
                if (version.startsWith('5')) {
                    for (const word of wordsInStatsv5) {
                        expect(data).toContain(word);
                    }
                } else {
                    for (const word of wordsInStatsv4) {
                        expect(data).toContain(word);
                    }
                }
                proc.kill();
                done();
                return;
            }
        });
    });
});
