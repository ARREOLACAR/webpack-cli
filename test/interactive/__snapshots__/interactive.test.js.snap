// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`--interactive flag should work with --watch 1`] = `
"Hash: e0f9eecf6c179bc310c7
Version: webpack 4.44.1
Time: 21ms
Built at: 09/11/2020 17:29:59
  Asset      Size  Chunks  Chunk Names
main.js  3.77 KiB    main  main
Entrypoint main = main.js
[./index.js] 27 bytes {main} [built]"
`;
